File: Laser_Takahashi.cpp
#include <bits/stdc++.h>
using namespace std;

#define endl '\n'
#define int long long

struct monster {
    int id;
    int x, y;
    long double angle;
};

void solve() {
    int n, q;
    cin >> n >> q;
    vector<monster> monsters(n);
    for (int i = 0; i < n; ++i) {
        monsters[i].id = i;
        cin >> monsters[i].x >> monsters[i].y;
        monsters[i].angle = atan2l(monsters[i].y, monsters[i].x);
    }
    sort(monsters.begin(), monsters.end(), [&](monster a, monster &b) {
        return a.angle < b.angle;
    });
    vector<int> cnt;
    vector<int> original_idx(n);
    cnt.emplace_back(1);
    original_idx[monsters[0].id] = 0;
    int now = 0;
    for (int i = 1; i < n; ++i) {
        if (monsters[i].angle == monsters[i - 1].angle) {
            cnt[now]++;
        } else {
            now++;
            cnt.emplace_back(1);
        }
        original_idx[monsters[i].id] = now;
    }
    vector<int> pre(cnt.size() + 1, 0);
    for (int i = 0; i < cnt.size(); ++i) {
        pre[i + 1] = pre[i] + cnt[i];
    }
    while (q--) {
        int left, right;
        cin >> left >> right;
        left--, right--;
        int original_left = original_idx[left], original_right = original_idx[right];
        if (original_left >= original_right) {
            cout << pre[original_left + 1] - pre[original_right] << endl;
        } else {
            cout << pre[original_left + 1] - pre[0] + pre[cnt.size()] - pre[original_right] << endl;
        }
    }
}

signed main() {
    cin.tie(nullptr)->sync_with_stdio(false);
    cout << fixed << setprecision(15);
    solve();
    return 0;
}

File: Laser_Takahashi.py
import sys
import math
input = lambda: sys.stdin.readline().strip()

n, q = map(int, input().split())
monsters = []
for i in range(n):
    x, y = map(int, input().split())
    monsters.append((math.atan2(y, x), i))
monsters.sort(key=lambda x: x[0])

cnt = []
original_idx = [0] * n

if n > 0:
    cnt.append(1)
    original_idx[monsters[0][1]] = 0
    now = 0
    
    for i in range(1, n):
        if monsters[i][0] == monsters[i-1][0]:
            cnt[now] += 1
        else:
            now += 1
            cnt.append(1)
        original_idx[monsters[i][1]] = now

pre = [0] * (len(cnt) + 1)
for i in range(len(cnt)):
    pre[i + 1] = pre[i] + cnt[i]

results = []
cnt_size = len(cnt)

for _ in range(q):
    left, right = map(int, input().split())
    left, right = left - 1, right - 1
    original_left = original_idx[left]
    original_right = original_idx[right]
    if original_left >= original_right:
        print(pre[original_left + 1] - pre[original_right])
    else:
        print(pre[original_left + 1] - pre[0] + pre[cnt_size] - pre[original_right])
